presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.metadata;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.operator.aggregation.InternalAggregationFunction;
+import com.facebook.presto.operator.scalar.ScalarFunctionImplementation;
+import com.facebook.presto.operator.window.WindowFunctionSupplier;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.BlockEncodingSerde;
+import com.facebook.presto.spi.function.OperatorType;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.facebook.presto.sql.SqlPathElement;
+import com.facebook.presto.sql.analyzer.FeaturesConfig;
+import com.facebook.presto.sql.analyzer.TypeSignatureProvider;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.spi.StandardErrorCode.FUNCTION_NOT_FOUND;
+import static com.facebook.presto.spi.StandardErrorCode.NOT_FOUND;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class FunctionManager
+{
+    private final FunctionNamespace globalFunctionNamespace;
+    private volatile Map<SqlPathElement, FunctionNamespace> dynamicFunctionNamespaces;
```

Comment:
Currently we only have a single subclass (`StaticFunctionNamespace`) of `FunctionNamespace`, and we use instances of  `StaticFunctionNamespace` for the `dynamicFunctionNamespaces` field, which is confusing. Looking at the rest of the patch, do you think we will need a `DynamicFunctionNamespace` class? If we won't, maybe we should make `FunctionNamespace` a class instead of an interface, and use it for both static/dynamic namespaces. Thoughts?

Reply:
Within this three diffs dynamicFunctionNamespaces are not really used. If we really don't want to introduce potentially unnecessary logic, I can change FunctionNamespace to be a class for now (and we can change that later when needed). We can remove the dynamicFunctionNamespaces private variable all together for now.